File: 2024/YounSangHo/Week7/2024-11-10-design-word-search-data-structure.js
class TrieNode {
  constructor() {
    this.children = Array(26).fill(null);
    this.endOfword = false;
  }
}

class WordDictionary {
  constructor() {
    this.root = new TrieNode();
  }

  /**
   * @param {string} c
   * @return {number}
   */
  getIndex(c) {
    return c.charCodeAt(0) - "a".charCodeAt(0);
  }

  /**
   * @param {string} word
   * @return {void}
   */
  addWord(word) {
    let cur = this.root;

    for (let c of word) {
      const idx = this.getIndex(c);
      if (cur.children[idx] === null) {
        cur.children[idx] = new TrieNode();
      }
      cur = cur.children[idx];
    }

    cur.endOfword = true;
  }

  /**
   * @param {string} word
   * @return {boolean}
   */
  search(word) {
    return this.dfs(word, 0, this.root);
  }

  dfs(word, j, root) {
    let cur = root;

    for (let i = j; i < word.length; i++) {
      const c = word[i];
      if (c === ".") {
        for (const child of cur.children) {
          if (child !== null && this.dfs(word, i + 1, child)) {
            return true;
          }
        }

        return false;
      } else {
        const idx = this.getIndex(c);
        if (cur.children[idx] === null) {
          return false;
        }
        cur = cur.children[idx];
      }
    }

    return cur.endOfword;
  }
}

File: 2024/Jooyoung/2024-11-10-DesignWordSearchDataStructure.py
class WordNode:
    def __init__(self) -> None:
        self.children = {}
        self.end_node = False


class WordDictionary:

    def __init__(self):
        self.root = WordNode()

    def addWord(self, word: str) -> None:
        cur = self.root

        for w in word:
            if w not in cur.children:
                cur.children[w] = WordNode()

            cur = cur.children[w]

        cur.end_node = True

    def search(self, word: str) -> bool:
        def dfs(j, root):
            cur = root

            for i in range(j, len(word)):
                if word[i] == ".":
                    for child in cur.children.values():
                        if dfs(i + 1, child):
                            return True
                    return False

                else:
                    if word[i] not in cur.children:
                        return False

                    cur = cur.children[word[i]]

            return cur.end_node

        return dfs(0, self.root)
